### DISCUSSION TITLE
@ObservableObject instead of @EnvironmentObject to init charts?

### DISCUSSION THREAD

**Discussion by juanjonol** [AppPear/ChartView#223] [Ideas]
Hi everyone! I just found this library and I like it a lot! I would like to contribute, but I have a few doubts about the current implementation.

One of the big changes of the new v2 version seems to be that [it allows to initialise a chart using an `@EnvironmentObject`](https://github.com/AppPear/ChartView/wiki/How-to-create-your-first-chart). This seems better than v1, where a chart requires static data and as such it won't automatically update when that data changes. But what I don't understand is, **why an `@EnvironmentObject` is used instead of an `@ObservableObject`?**

As I see it, `@EnvironmentObject` is better suited for pieces of data that are common to a whole app (current `colorScheme`, font...), but in this case we could have hundred of charts with different data. Also, because the data isn't set as part of the `init()`, its very easy to forget setting it, which could make the whole app crash (if `.enviromentObject()` isn't called on any parent View)[^1], or even worse, it could make a chart use unexpected data (if a parent View calls `.enviromentObject()` for an unrelated chart).

Keeping all this in mind, I think it would make more sense to:

- Keep an `init()` similar to the one on v1, that forces setting the actual data that will be used by the chart (for example, ` BarChartView(data:)`), but with the requirement that this data is an `@ObservableObject`.
- Keep View Modifiers like `.chartStyle()` to modify the View on a SwiftUI-idiomatic way, but without using `@EnvironmentObject` internally.
- Finally, maybe a few `@EnvironmentObjects` could be added, for example a `defaultChartStyle` in case no `chartStyle` is set during the chart's initialisation.

Is there anything I'm missing? Would this proposal make sense?

Thank you!

[^1]: This literally happened to me, before I found the v2 beta documentation.

**Comment by kushalpagolu** [comment]
Best Regards,
P.Kushal,
3056997282.
